### PR TITLE
[inductor] masked(): emit explicit where when the load is served from the CSE store cache

### LINKED ISSUE #189895: [inductor] Silent wrong results: constant_pad_nd fused into a split reduction drops the pad's boundary select (corrupts flash-attn cu_seqlens)
### 🐛 Describe the bug

`torch.compile` produces silently wrong results for the standard "unpad" pattern `F.pad(x.sum(...), (1, 0))` when Inductor implements the sum as a **split reduction**. The `constant_pad_nd` gets fused into the reduction's second-stage kernel and its boundary select (`where(idx - 1 >= 0, value, 0)`) is dropped: the leading pad element receives the reduction result instead of the fill value `0`.

```python
import torch
import torch.nn.functional as F

def cu_seqlens(mask):
    lengths = mask.sum(dim=-1, dtype=torch.int32)
    return F.pad(torch.cumsum(lengths, dim=0, dtype=torch.int32), (1, 0))

mask = torch.ones(1, 16384, device="cuda", dtype=torch.bool)

print("eager   :", cu_seqlens(mask).tolist())
print("compiled:", torch.compile(cu_seqlens, fullgraph=True)(mask).tolist())
```

Output (use a fresh `TORCHINDUCTOR_CACHE_DIR` — see "Trigger conditions" below):

```
eager   : [0, 16384]
compiled: [16384, 16384]   # expected [0, 16384]
```

The `cumsum` is not required — `F.pad(mask.sum(dim=-1, dtype=torch.int32), (1, 0))` alone reproduces it. Same result for int32/int64/float dtypes, with/without `dynamic=True`, and with/without `inference_mode`.

### Generated kernel

With `TORCH_LOGS=output_code`, the second stage of the split sum is fused with the pad into `triton_per_fused_constant_pad_nd_cumsum_sum_1`. The pad's boundary mask (`tmp7`) is computed but **never used**, and the reduction result is stored unconditionally to every output position, including the pad slot:

```python
tmp0 = tl.load(in_ptr0 + (r0_0), r0_mask, eviction_policy='evict_first', other=0.0)  # partial sums from stage 1
tmp1 = tl.broadcast_to(tmp0, [XBLOCK, R0_BLOCK])
tmp3 = tl.where(r0_mask, tmp1, 0)
tmp4 = tl.sum(tmp3, 1)[:, None].to(tl.int32)
tmp5 = ((-1) + r0_0).to(tl.int32)
tmp6 = tl.full([1, 1], 0, tl.int64)
tmp7 = tmp5 >= tmp6                                                                  # pad boundary mask: DEAD CODE
tl.store(out_ptr1 + (tl.broadcast_to(r0_0, [XBLOCK, R0_BLOCK])), tmp4, r0_mask)
```

### Trigger conditions

All verified empirically; because cached artifacts affect the outcome (see condition 4), each check below was run in an isolated fresh process with its own fresh `TORCHINDUCTOR_CACHE_DIR`.

1. **The reduction must be split.** Large reduced dim (16384 elements → 2 partials of 8192) reproduces; 4096 does not. `torch._inductor.config.split_reductions = False` makes the result correct, pinning the bug to the split-reduction + pad fusion.
2. **Batch size 1.** The pad output must have exactly 2 elements to be mis-fused with the 2-partial second stage; `mask` shapes `(2, L)` / `(4, L)` generate correct code.

3. **With `dynamic=True`, the first call's size hint decides — and poisons all sizes.** The split decision is baked into the compiled artifact:
    | first call | then called with | result |
    |---|---|---|
    | L=16384 (split) | L=16384 | ❌ `[16384, 16384]` |
    | L=16384 (split) | L=4096 | ❌ `[4096, 4096]` — wrong for **all** sizes afterwards |
    | L=4096 (no split) | L=16384 | ✅ correct for all sizes |

4. **On-disk caches propagate the bad artifact across processes.** A fresh process that reuses a `TORCHINDUCTOR_CACHE_DIR` populated by a large-hint compile computes `[4096, 4096]` even though it only ever calls L=4096. Combined with (3), this makes the bug appear flaky/heisenbug-like in real workloads and hard to reproduce unless the cache dir is controlled.

### Real-world impact

This exact pattern is the standard `_get_unpad_data` / `unpad_input` prologue used by HF transformers (`modeling_flash_attention_utils.py`) and flash-attn (`bert_padding.py`) to build `cu_seqlens` for `flash_attn_varlen_func`:
https://github.com/huggingface/transformers/blob/ab1771c9e42891d893189978a8009426d70b4688/src/transformers/modeling_flash_attention_utils.py#L321
https://github.com/huggingface/transformers/blob/ab1771c9e42891d893189978a8009426d70b4688/src/transformers/modeling_flash_attention_utils.py#L371
https://github.com/Dao-AILab/flash-attention/blob/2402cb0bed7a2185cb9ddbe88fb998656cf73066/flash_attn/bert_padding.py#L116
https://github.com/Dao-AILab/flash-attention/blob/2402cb0bed7a2185cb9ddbe88fb998656cf73066/flash_attn/bert_padding.py#L190

With the corrupted `cu_seqlens = [N, N]`, the varlen kernel treats the batch as zero-length and **never writes its output buffer**, so the model returns uninitialized memory — intermittent garbage/NaN that changes between iterations. Batch size 1 with long sequences is the common single-sample inference case for diffusion transformers and LLM prefill. The corruption is hard to trace back: the attention inputs q/k/v are all bit-exact correct; only the tiny `cu_seqlens` tensor is wrong.

Workaround: build the offsets with `torch.cat` instead of `F.pad`, which compiles correctly under both codegen paths (verified with both a split-triggering and a non-split first-call size hint):

```python
c = torch.cumsum(lengths, dim=0, dtype=torch.int32)
cu_seqlens = torch.cat([c.new_zeros(1), c])
```

### Versions

```
Collecting environment information...
PyTorch version: 2.13.0+cu132
Is debug build: False
CUDA used to build PyTorch: 13.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.4 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04.1) 13.3.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.39

Python version: 3.11.15 | packaged by conda-forge | (main, Jun 11 2026, 03:34:02) [GCC 14.3.0] (64-bit runtime)
Python platform: Linux-6.18.33.2-microsoft-standard-WSL2-x86_64-with-glibc2.39
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to:
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 4090
Nvidia driver version: Could not collect
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: False
Caching allocator config: N/A

CPU:
Architecture:                            x86_64
CPU op-mode(s):                          32-bit, 64-bit
Address sizes:                           39 bits physical, 48 bits virtual
Byte Order:                              Little Endian
CPU(s):                                  24
On-line CPU(s) list:                     0-23
Vendor ID:                               GenuineIntel
Model name:                              13th Gen Intel(R) Core(TM) i7-13700KF
CPU family:                              6
Model:                                   183
Thread(s) per core:                      2
Core(s) per socket:                      12
Socket(s):                               1
Stepping:                                1
BogoMIPS:                                6835.19
Flags:                                   fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc rep_good nopl xtopology tsc_reliable nonstop_tsc cpuid tsc_known_freq pni pclmulqdq vmx ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves avx_vnni vnmi umip waitpkg gfni vaes vpclmulqdq rdpid movdiri movdir64b fsrm md_clear serialize ibt flush_l1d arch_capabilities
Virtualization:                          VT-x
Hypervisor vendor:                       Microsoft
Virtualization type:                     full
L1d cache:                               576 KiB (12 instances)
L1i cache:                               384 KiB (12 instances)
L2 cache:                                24 MiB (12 instances)
L3 cache:                                30 MiB (1 instance)
NUMA node(s):                            1
NUMA node0 CPU(s):                       0-23
Vulnerability Gather data sampling:      Not affected
Vulnerability Ghostwrite:                Not affected
Vulnerability Indirect target selection: Not affected
Vulnerability Itlb multihit:             Not affected
Vulnerability L1tf:                      Not affected
Vulnerability Mds:                       Not affected
Vulnerability Meltdown:                  Not affected
Vulnerability Mmio stale data:           Not affected
Vulnerability Old microcode:             Not affected
Vulnerability Reg file data sampling:    Mitigation; Clear Register File
Vulnerability Retbleed:                  Mitigation; Enhanced IBRS
Vulnerability Spec rstack overflow:      Not affected
Vulnerability Spec store bypass:         Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:                Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:                Mitigation; Enhanced / Automatic IBRS; IBPB conditional; PBRSB-eIBRS SW sequence; BHI BHI_DIS_S
Vulnerability Srbds:                     Not affected
Vulnerability Tsa:                       Not affected
Vulnerability Tsx async abort:           Not affected
Vulnerability Vmscape:                   Not affected

Versions of relevant libraries:
[pip3] flash_attn==2.8.3+cu132torch2.13
[pip3] numpy==2.4.4
[pip3] nvidia-cublas==13.4.0.1
[pip3] nvidia-cuda-cupti==13.2.75
[pip3] nvidia-cuda-nvrtc==13.2.78
[pip3] nvidia-cuda-runtime==13.2.75
[pip3] nvidia-cudnn-cu13==9.20.0.48
[pip3] nvidia-cufft==12.2.0.46
[pip3] nvidia-curand==10.4.2.55
[pip3] nvidia-cusolver==12.2.0.1
[pip3] nvidia-cusparse==12.7.10.1
[pip3] nvidia-cusparselt-cu13==0.8.1
[pip3] nvidia-nccl-cu13==2.29.7
[pip3] nvidia-nvjitlink==13.2.78
[pip3] nvidia-nvtx==13.2.75
[pip3] torch==2.13.0+cu132
[pip3] torchvision==0.28.0+cu132
[pip3] triton==3.7.1
[conda] flash-attn                 2.8.3+cu132torch2.13  pypi_0                pypi
[conda] numpy                      2.4.4                 pypi_0                pypi
[conda] nvidia-cublas              13.4.0.1              pypi_0                pypi
[conda] nvidia-cuda-cupti          13.2.75               pypi_0                pypi
[conda] nvidia-cuda-nvrtc          13.2.78               pypi_0                pypi
[conda] nvidia-cuda-runtime        13.2.75               pypi_0                pypi
[conda] nvidia-cudnn-cu13          9.20.0.48             pypi_0                pypi
[conda] nvidia-cufft               12.2.0.46             pypi_0                pypi
[conda] nvidia-curand              10.4.2.55             pypi_0                pypi
[conda] nvidia-cusolver            12.2.0.1              pypi_0                pypi
[conda] nvidia-cusparse            12.7.10.1             pypi_0                pypi
[conda] nvidia-cusparselt-cu13     0.8.1                 pypi_0                pypi
[conda] nvidia-nccl-cu13           2.29.7                pypi_0                pypi
[conda] nvidia-nvjitlink           13.2.78               pypi_0                pypi
[conda] nvidia-nvtx                13.2.75               pypi_0                pypi
[conda] torch                      2.13.0+cu132          pypi_0                pypi
[conda] torchvision                0.28.0+cu132          pypi_0                pypi
[conda] triton                     3.7.1                 pypi_0                pypi
```

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo

### PR DESCRIPTION
**Pull request by cynricfu**
Fixes #189895

`TritonKernelOverrides.masked()` skips the explicit `tl.where` when the masked body is a plain load, pushing the fill value into `tl.load(..., mask=..., other=...)` instead. However, when the loaded buffer's producer is fused into the same kernel, `CSEProxy.load()` returns the cached CSE variable from `cse.store_cache` and emits no `tl.load` at all — the mask and `other` are silently dropped, and `masked()` returns the unguarded value.

Observable effect: `F.pad(x.sum(...), (1, 0))` with a split reduction returns `[N, N]` instead of `[0, N]`. This is the standard `cu_seqlens` unpad prologue used by HF transformers and flash-attn, so under torch.compile the corrupted offsets make `flash_attn_varlen_func` treat the batch as zero-length and return uninitialized memory (see issue for details and controls).

The fix makes the `need_where` scan treat a load served from the store cache as requiring the explicit `where`. Loads that actually lower to `tl.load` keep the existing masked-load fast path.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Review by eellison — APPROVED** [review]
